### DISCUSSION TITLE
MeshCentral behind Nginx Proxy Manager

### DISCUSSION THREAD

**Discussion by adimate71** [Ylianst/MeshCentral#6144] [General]
Hi,

I would like to deploy MeshCentral behind the Nginx Proxy Manager. 
The Nginx Proxy Manager Docker container is installed on a Raspberry Pi with IP 192.168.1.50. 
MeshCentral is installed on another Raspberry Pi with IP 192.168.1.40 (Ubuntu). 
On Nginx Proxy Manager the Let's Encrypt certificate is configured to use DNS challenge (Cloudflare API), so only port 443 is forwarded to 192.168.1.50 in my router.

I configured the Nginx Proxy Manager, and MeshCentral is available externally, but the agents do not communicate with MeshCentral. All agents are installed on remote devices. I tried to add the following settings in config.json, but MeshCentral stopped working (bad gateway error):
"certUrl": "https://mesh.domain.com:443"
"TLSOffload": "192.168.1.50"

Could you please help me to modify the config.json to allow the external agents to communicate with MC?

Please see below my current json configuration.
```
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "_cert": "mesh.domain.com",
    "WANonly": true,
    "_LANonly": true,
    "_sessionKey": "MyReallySecretPassword1",
   "TLSOffload": "192.168.1.50",
    "port": 443,
    "_aliasPort": 443,
    "redirPort": 80,
    "_redirAliasPort": 80
  },
  "domains": {
    "": {
      "_title": "domain",
      "_title2": "Servername",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": “username@gmail.com",
    "names": "mesh.domain.com",
    "certUrl": "https://mesh.domain.com:443",
    "skipChallengeVerification": true,
    "production": true
  }
}
```
Thank you

**Comment by si458** [comment]
`certUrl` should be placed under the `domains` then `""` section NOT LETSENCRYPT
p.s dont forget to restart meshcentral after u change ur config.json

**Reply by adimate71** [reply]
I made the correction and restarted MC but still getting the same error "502 Bad Gateway". 
If I comment out "TLSOffload" and restart the server, I can access MC. 
Could you please suggest any other settings you would recommend?
Thank you

**Reply by si458** [reply]
the 'Bad Gateway' is an error from nginx saying it couldnt access your meshcentral server,
so check the logs/console output from meshcentral is the is any errors when it starts up

how is meshcentral also setup? docker? npm?

also you should set `"_cert": ""` to `"cert": "DOMAINHERE"`

**Comment by adimate71** [comment]
Meshcentral is installed with NPM.
I made the change "_cert": "" to "cert": "mesh.domain.com" - same error if I uncomment TLSOffload.
I am not sure how to collect error logs - I used the below command to debug.


root@server:/home/mc# node node_modules/meshcentral --debug
MAIN: Core module windows-amt is 646343 bytes.
MAIN: Core module linux-amt is 592266 bytes.
MAIN: Core module linux-noamt is 491730 bytes.
MAIN: Core module windows-recovery is 137824 bytes.
MAIN: Core module linux-recovery is 82212 bytes.
MAIN: Core module windows-agentrecovery is 61977 bytes.
MAIN: Core module linux-agentrecovery is 6365 bytes.
MAIN: Core module windows-tiny is 6305 bytes.
MAIN: Core module linux-tiny is 6305 bytes.
MeshCentral HTTP redirection server running on port 81.
AUTHLOG: Server listening on 0.0.0.0 port 81.
CERT: LE: Getting certs from local store (Production)
WARNING: Redirection web server must be active on port 80 for Let's Encrypt to work.
CERT: LE: WARNING: Redirection web server must be active on port 80 for Let's Encrypt to work.
MeshCentral v1.1.24, WAN mode.
DISPATCH: AddEventDispatch [ 'server-shareremove' ]
CERT: loadCertificate() - Loading certificate from mesh.domain.com:443, Hostname: mesh.domain.com...
DISPATCH: AddEventDispatch [ '*' ]
DISPATCH: DispatchEvent [ '*' ]
MAIN: Server started
ERR: ERROR: MeshCentral Intel(R) AMT server port 4433 is not available. Check if the MeshCentral is already running.
HTTPS: Server listening on 0.0.0.0 port 444.
MeshCentral HTTPS server running on mesh.domain.com:444.
CERT: loadCertificate() - TLS connected, got certificate.
Loaded web certificate from "https://mesh.domain.com:443", host: "mesh.domain.com"
  SHA384 cert hash: 85743333333

**Reply by si458** [reply]
you have multiple meshcentrals running on your server.
this is because its running on port 81 and 444, even tho i specified 80 and 443 in your config.json
have you tried simply restarting ur whole server/vm?

**Comment by adimate71** [comment]
I rebooted the server - same error.
I would like to mention that if comment out "TLSOffload" and I forward the port 443 to the MC server's IP (192.168.1.40) instead of Ngin Proxy Manager's IP, everything works fine, the agents are present in MC.

**Reply by si458** [reply]
What have u set as the url inside nginx proxy manager?
If u enable tlsoffload (which u should do as ur using nginx)
You should have it set to `http://meshcentralip:443`
(don't use https! If u have tlsoffload set)

Share a screenshot of ur nginx config page

**Comment by bolavefasfas** [comment]
I am using with npm 
I was facing same issue I fixed it using disabling cloudflare proxy from dns in cloudflare portal 

watch aussomeopensource video on youtube he literally spoonfeeds complete installation.
and in npm I had to set https or else it wont work and my origin server certificate does not work  (only for meshcentral for other it does ) so I used lets encrypt everything works now.

**Comment by JoSoG89** [comment]
im using the same proxy and this is my config.json:

my case:
Nginx proxy manager:
https ip 192.168.1.10 port to listen 30022 in your case will be port 443
TLSOffload is the gateway (your router)

all the info in "_letsencrypt" will be ignored, the cert will be generated by nginx and the meshcentral will request the cert to nginx


{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "settings": {
    "plugins":{"enabled": true},
    "postgres": {"some info"},
    "cert": "somedomain.com",
    "_WANonly": true,
    "_LANonly": true,
    "sessionKey": "somekey",
    "port": 443,
    "_aliasPort": 443,
    "redirPort": 80,
    "_redirAliasPort": 80,
    "AgentPong": 300,
    "TLSOffload": "192.168.1.1",
    "SelfUpdate": false,
    "AllowFraming": false,
    "WebRTC": false
  },
  "domains": {
    "": {
      "_title": "some title",
      "_title2": "This is a private server",
      "minify": true,
      "NewAccounts": true,
      "localSessionRecording": false,
      "_userNameIsEmail": true,
      "certUrl": "https://192.168.1.10:30022" 
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before>",
    "_email": "myemail@mydomain.com",
    "_names": "myserver.mydomain.com",
    "production": false
  }
}


This is the server taking the cert from proxy
![image](https://github.com/Ylianst/MeshCentral/assets/106213922/d041253a-3e69-4735-829d-ec5ab8e0cc2c)

**Comment by ficoreyes** [comment]
change "certUrl": "https://192.168.1.10:30022/" to your domain (example)  "certUrl": "https://meshcentral.yourdomain.com:443"
